### PR TITLE
Add gpu implementation of ResourceSparseApplyAdadelta Op for issue #2314 

### PR DESCRIPTION
**Pull request by irvineoy**
# This commit is related to issue #2314 and #30537.
### When I want to use Adadelta Optimizer in GPU, I encounted the following error：
Error environment info:
- Ubuntu 16
- Python 3.6
- Tensorflow 2.3
- CUDA 10.1
- GPU Tesla V100
- GPU Driver 396.37

```
File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/execute.py", line 60, in quick_execute  
inputs, attrs, num_outputs)  
tensorflow.python.framework.errors\_impl.InvalidArgumentError: Cannot assign a device for operation Adadelta/Adadelta/update/Unique: Could not satisfy explicit device specification '' because the node {{colocation\_node Adadelta/Adadelta/update/Unique}} was colocated with a group of nodes that required incompatible device '/job:localhost/replica:0/task:0/device:GPU:0'. All available devices \[/job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:XLA\_CPU:0, /job:localhost/replica:0/task:0/device:XLA\_GPU:0, /job:localhost/replica:0/task:0/device:GPU:0\].  
Colocation Debug Info:  
Colocation group had the following types and supported devices:  
Root Member(assigned\_device\_name\_index\_=2 requested\_device\_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned\_device\_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource\_device\_name_='/job:localhost/replica:0/task:0/device:GPU:0' supported\_device\_types_=\[CPU\] possible\_devices\_=\[\]  
Identity: GPU CPU XLA\_CPU XLA\_GPU  
ReadVariableOp: GPU CPU XLA\_CPU XLA\_GPU  
ResourceSparseApplyAdadelta: CPU  
Unique: GPU CPU  
Shape: GPU CPU XLA\_CPU XLA\_GPU  
\_Arg: GPU CPU XLA\_CPU XLA_GPU  
Const: GPU CPU XLA\_CPU XLA\_GPU  
StridedSlice: GPU CPU XLA\_CPU XLA\_GPU  
UnsortedSegmentSum: GPU CPU XLA\_CPU XLA\_GPU

Colocation members, user-requested devices, and framework assigned devices, if any:  
wide\_deep\_8564 (_Arg) framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0  
adadelta\_adadelta\_update\_resourcesparseapplyadadelta\_accum (_Arg) framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0  
adadelta\_adadelta\_update\_resourcesparseapplyadadelta\_accum\_update (\_Arg) framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0  
Adadelta/Adadelta/update/Unique (Unique)  
Adadelta/Adadelta/update/Shape (Shape)  
Adadelta/Adadelta/update/strided_slice/stack (Const)  
Adadelta/Adadelta/update/strided\_slice/stack\_1 (Const)  
Adadelta/Adadelta/update/strided\_slice/stack\_2 (Const)  
Adadelta/Adadelta/update/strided_slice (StridedSlice)  
Adadelta/Adadelta/update/UnsortedSegmentSum (UnsortedSegmentSum)  
Adadelta/Adadelta/update/ResourceSparseApplyAdadelta (ResourceSparseApplyAdadelta)  
wide\_deep/StatefulPartitionedCall/embedding\_lookup\_sparse\_30/embedding\_lookup\_sparse/embedding_lookup/ReadVariableOp (ReadVariableOp)  
Func/wide\_deep/StatefulPartitionedCall/input/\_148 (Identity) /job:localhost/replica:0/task:0/device:GPU:0

     [[{{node Adadelta/Adadelta/update/Unique}}]] [Op:__inference_train_step_9850]
```

To fix this, I wrote the GPU implementation for ResourceSparseApplyAdadelta Op and tested in r2.3 branch.
Test Environment info:
- Docker Image: tensorflow/tensorflow:nightly-custom-op-gpu-ubuntu16
- Ubuntu 16
- Python 3.6
- Tensorflow r2.3
- CUDA 10.1
- cuDNN 7
- Cuda Compute Capabilities 7.0 (Tesla V100 Driver 396.37)
- GCC 7
- bazel 3.1.0

### CODE REVIEW

File: tensorflow/core/kernels/training_ops.cc [hunk 1]
```diff
@@ -1003,32 +1037,41 @@ REGISTER_KERNELS(GPU, complex128);
...
     const bool sparse = true;
-    auto locks = MaybeLockVariableInputMutexesInOrder<CPUDevice, T>(
-        ctx, use_exclusive_lock_, sparse, {0, 1, 2});
-    DoCompute(ctx);
+    mutex* mu = GetTrainingVariableMutex<Device, T>(ctx, 0, sparse, &var);
+    core::ScopedUnref scoped_unref(var);
+    // mu_accum is actually the same mutex as mu_var since currently we use a
+    // global mutex.
+    //
+    // mutex* mu_accum = ctx->input_ref_mutex(1);
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:1055]: How about the mutex for `accum_update`?

Also, I don't really understand the comment; if `accum` and `var` are different ref variables then won't they have different mutexes?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/training_ops.cc [hunk 2]
```diff
@@ -581,6 +581,40 @@ struct SparseApplyKerasMomentum<CPUDevice, T, Tindex> {
...
+                  typename TTypes<T>::ConstScalar rho,
+                  typename TTypes<T>::ConstScalar epsilon,
+                  typename TTypes<T>::ConstMatrix grad,
+                  typename TTypes<Tindex>::ConstFlat indices){
+    const Tindex N = static_cast<Tindex>(indices.size());
+    for (Tindex i = 0; i < N; i++) {
+      const Tindex index = indices(i);
+      auto accum_ = accum.template chip<0>(index);
+      auto accum_update_ = accum_update.template chip<0>(index);
+      auto grad_ = grad.template chip<0>(i);
```
[the remark this answers is not in the document]
Comment (irvineoy) [reply tensorflow/core/kernels/training_ops.cc]: OK, I will change the style

File: tensorflow/core/kernels/training_ops.cc (hunk 1 above)
Comment (irvineoy) [reply tensorflow/core/kernels/training_ops.cc]: I have changed the logic of this part.

**Comment by irvineoy** [comment]
It's my first time to create a pull request. About that Intel oneDNN unit tests, It seems some errors are not related to the part I modified. Is that a problem?

**Comment by irvineoy** [comment]
I have modified the code base on the comment, could you review it? @sanjoy

**Comment by pawel-polyai** [comment]
When this PR will be reviewed again? This is really blocking training in distributed manner for sparse tensors (common in NLP).

File: tensorflow/core/kernels/training_ops_gpu.cu.cc [hunk 3]
```diff
@@ -637,6 +637,69 @@ struct SparseApplyKerasMomentum<GPUDevice, T, Tindex> {
   }
 };
 
+template <typename T, typename Tindex>
+__global__ __launch_bounds__(1024) void SparseApplyAdadeltaKernel(
+    T* var, T* accum, T* accum_update, const T* lr, const T* rho,
+    const T* epsilon, const T* grad, const Tindex* indices, Tindex param_rows,
+    Tindex updates_size, Tindex indices_size) {
+  Tindex col_size = updates_size / indices_size;
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops_gpu.cu.cc:645]: Do you mean `row_size`?

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@irvineoy  Can you please check @sanjoy's comments and keep us posted ? Thanks!

File: tensorflow/core/kernels/training_ops_gpu.cu.cc (hunk 3 above)
Comment (irvineoy) [reply tensorflow/core/kernels/training_ops_gpu.cu.cc:645]: Based on other kernel implementation convention, like `SparseApplyAdagradKernel`, `SparseApplyFtrlKernel` and etc. I named it `col_size`. Do I need to change it?

**Comment by gbaned** [comment]
@irvineoy Any update on this PR? Please. Thanks!

**Comment by irvineoy** [comment]
> @irvineoy Any update on this PR? Please. Thanks!

I am waiting for sanjoy's reply. 
@sanjoy mentioned a variable name. I check the source code again. And I found it's quite normal to name that variable as `col_size`. So I wonder if it's necessary to change it to another name.

**Comment by gbaned** [comment]
@sanjoy  Can you please take a look on above comments from @irvineoy. Thanks!

**Review by sanjoy — APPROVED** [review]

File: tensorflow/core/kernels/training_ops.cc [hunk 4]
```diff
@@ -1157,33 +1185,32 @@ REGISTER_KERNELS(GPU, complex128);
...
-template <typename T, typename Tindex>
+template <typename T, typename Device, typename Tindex>
 class SparseApplyAdadeltaOp : public OpKernel {
  public:
   explicit SparseApplyAdadeltaOp(OpKernelConstruction* ctx) : OpKernel(ctx) {
     OP_REQUIRES_OK(ctx, ctx->GetAttr("use_locking", &use_exclusive_lock_));
   }
 
-  void Compute(OpKernelContext* ctx) override {
+  void Compute(OpKernelContext* ctx) override TF_NO_THREAD_SAFETY_ANALYSIS {
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:1195]: Please add a comment on why `TF_NO_THREAD_SAFETY_ANALYSIS` is needed.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/training_ops.cc (hunk 4 above)
Comment (irvineoy) [reply tensorflow/core/kernels/training_ops.cc:1196]: Done

Comment (sanjoy) [reply tensorflow/core/kernels/training_ops.cc:1196]: Can you please add a bit more detail on what's the race and why it is benign?

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Comment by irvineoy** [comment]
At the very beginning, I noticed almost all the `SparseXXXOp` classes had such macro, so I added it to SparseApplyAdadeltaOp. As for the detailed reason, firstly I checked all the commit logs of `tensorflow/core/kernels/training_ops.cc`. I found such macro was firstly introduced by Dr. Manjunath Kudlur in the Initial commit of TensorFlow for `SparseApplyMomentumOp` and `SparseApplyAdagradOp` classes without explicitly explaining why. Then I went to the llvm's thread safety analysis document and found there is [one of the limitations](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-conditionally-held-locks) may be related to this issue because inside the function `MaybeLockVariableInputMutexesInOrder` there is a conditional expression of `sparse` variable. And if you consider there is no need for this macro, I can remove it.

**Comment by xiazhiyi99** [comment]
Could please you speed up this review process? We also encountered this problem and it is quite annoying. @sanjoy

**Review by sanjoy — APPROVED** [review]

**Comment by irvineoy** [comment]
Hi Sanjay, It seems two CI are failing. Do I need to rebase my branch?

**Comment by sanjoy** [comment]
> Hi Sanjay, It seems two CI are failing. Do I need to rebase my branch?

I'm merging it manually, no action required from you for now.

**Pull request merged**